### DISCUSSION TITLE
Go back/forward

### DISCUSSION THREAD

**Discussion by reintroducing** [TanStack/router#181] [Q&A]
In react-router you have the ability to do `history.goBack`. I have combed through the docs and have not found a way to do this in react-location.  What is the best way to accomplish this?

**Comment by reintroducing** [comment]
Of course I found my answer right after posting this. `useLocation` exposes `history` which you can then call `back` on. @tannerlinsley the `useLocation` hook is not listed in the API docs so I wasn't aware of its existence until I found a different issue where you mentioned it.

**Reply by altmshfkgudtjr** [reply]
Thank you!
I bought my precious time.

I used like this.
```typescript
import { useLocation } from '@tanstack/react-location';
```

```typescript
const location = useLocation();
const onBack = () => location.history.back();
```

**Comment by iamolegga** [comment]
Late 2023 answer:

```ts
const router = useRouter();
const onBack = () => router.history.back();
```

**Reply by well1791** [reply]
Great!! Now how do I do that using `<Link>` component?

**Reply by bakesteve** [reply]
anyone done this? using history.go(-1) works per below, but be nicer to use a link, but cant see a way that  `const { history } = useRouter()` can pass a usable value to `<Link to={...}>..Back</Link>`

**Reply by herbievine** [reply]
any updates on how this can be achieved? @schiller-manuel

**Comment by rudzky** [comment]
Any update on this? I would like to able to navigate back only in the **scope of app routing** and don't go back to the page I was before I came to **this** app.

**Reply by fabiobaser** [reply]
I have the same issue. You can look into the history and dynamically route to different point.

**Comment by ento-io** [comment]
Such a basic thing should be in the docs

**Comment by AbdelhakimMhioul** [comment]
I used it like this :

```ts
 const { history } = useRouter()
 
 onClick={() => history.go(-1)}
```

**Comment by ghost** [comment]
None of the proposed solutions allow app back/forward navigation without a page reload. Would be very nice to have this feature.

**Comment by igor9silva** [comment]
Anyway to determine if there is something beyond the current route on the stack?
I want to dynamically enable/disable a "forward" button.

**Reply by igor9silva** [reply]
@tannerlinsley @SeanCassiere sorry to bother, I've searched everywhere. Only way I could achieve that was through subscribing to router events, which sounds hacky. Is there a better way?

**Reply by schiller-manuel** [reply]
using router events is not hacky.
however, we are building this in https://github.com/TanStack/router/pull/3017

**Reply by igor9silva** [reply]
thank you 🙏
Would love a `canGoForward` also

**Reply by schiller-manuel** [reply]
released now!

**Reply by SalahAdDin** [reply]
> released now!

what is it for?

**Reply by TrawenCZ** [reply]
For those looking for `canGoForward` alternative, this works for me, even though I'm not quite sure if I'm not accessing some internal stuff I shouldn't.
```tsx
  const router = useRouter();

  const canGoForward = router.history.location.state.__TSR_index < router.history.length - 1;
```

**Comment by zngly-vlad** [comment]
With the new release of `useCanGoBack`, it would be nice it if there was a way to specify the direction in the navigation.

```ts
// something like could be added to specify that we are going backwards, 
// but the "to" route is still provided in case canGoBack is false

// given the current route is /users/$userId
<Link to="/users" direction="back" />
```

Currently I am wrapping the `navigate` hook and `Link`component to use `router.history.go(-1)` conditionally based on the direction and if it can go back.

This would be really helpful in order to preserve url params and scroll-restoration

**Comment by primado** [comment]
I also faced this same issue, this is the 2nd project I'm working on that uses Tanstack router. I also combed docs  but I didn't see it listed in the docs. So instead of using a LLM to find I said lemme go with the "old school" way of doing it. I Googled and here am I.

@tannerlinsley Pls the Tanstack Router needs to be updated like how the React Query Docs is. Or at least show how these functionalities work. Thank you and thanks for all the good things you're doing for the community.

**Reply by schiller-manuel** [reply]
what exactly is missing?

**Reply by primado** [reply]
I meant, how to use the hooks in Tanstack Router. Because I have been going through the docs to find something like this but I couldn't see any, until I Googled.

```JSX
const router = useRouter();
const onBack = () => router.history.back();
```

**Reply by schiller-manuel** [reply]
can you contribute to the docs?

**Reply by primado** [reply]
> can you contribute to the docs?

I would love to do it later in the future. I will reach out to you when I'm ready.

**Comment by tomredman** [comment]
This is what I'm doing in order to use a `<Link />` to go back via `router.history.back()':

```typescript
  const router = useRouter();
  const onBack = () => router.history.back();

  <Link
    to={"/"}
    onClick={() => {
      onBack();
      return false;
    }}
  >
    <ArrowLeft />
    Back
  </Link>
```

Returning "false" from onClick() in an `<a>` tag (which `<Link />` is) surpresses the browsers default behavior. So here I'm effectively ignoring the `to` parameter and instead running onBack().

It works, with one little weird UX glitch that the URL flashes... _something_... very quickly on navigating back. I assume it has to do with goBack() racing with `return false`.

**Reply by deshazer** [reply]
This is a nice hack! You can also prevent the flash that you are referring to by using the event in your `onClick` function and doing `event.preventDefault()`

```
<Link
  to="/"
  onClick={(e) => {
    e.preventDefault()
    router.history.back()
    return false
  }}
>Back</Link>
```

**Reply by MPiland** [reply]
Serious question from someone still learning, why wouldn't you just use a button at this point? If you're ignoring the to portion, doesn't that render the component unnecessary and just adds an extra import?

**Reply by tomredman** [reply]
@MPiland now I'm trying to think back why I wanted to use a <Link> here – and I think it's because navigating with anything other than <Link> causes a full reload, and I wanted to keep that back functionality seamless.

**Comment by adamthewilliam** [comment]
https://tanstack.com/router/latest/docs/framework/react/api/router/useCanGoBack#showing-a-back-button

**Reply by adamthewilliam** [reply]
I wish there was a `useCanGoForward`

**Comment by spyrosbazios** [comment]
https://tanstack.com/router/latest/docs/framework/react/guide/navigation#special-relative-paths--and- 

Exmple from the docs : 

```export const Route = createFileRoute('/posts/$postId')({
  component: PostComponent,
})

function PostComponent() {
  return (
    <div>
      <Link to=".">Reload the current route of /posts/$postId</Link>
      <Link to="..">Navigate back to /posts</Link>
      // the below are all equivalent
      <Link to="/posts">Navigate back to /posts</Link>
      <Link from="/posts" to=".">
        Navigate back to /posts
      </Link>
      // the below are all equivalent
      <Link to="/">Navigate to root</Link>
      <Link from="/posts" to="..">
        Navigate to root
      </Link>
    </div>
  )
}```

**Comment by em843** [comment]
Using router.history.back() instead of Link or navigate means that I can't use options like `{resetScroll: false}`. It would be great if navigate and Link supported this.
